Filter extension

leewaygroups opened the pull request:
Hello,
I followed the discussion about one of the open issues:*\* "Exclude files #27"*\* and thought it would be nice to help out. Here is a pull request.

**Action:** Extended the filter param to also take either a callback or an array ( strings) of names of files to be excluded. Added few lines of test in the test.js file.

lvegerano commented:
@dougwilson @felixge is this change going to be merged.

dougwilson commented:
Hi @lvegerano, 1 day over a weekend is not enough time for me to look at the pull request. Please be patient.

lvegerano commented:
@dougwilson my apologies I did not payed attention to the day :+1:

holm commented:
Having the filter function would be pretty convenient for us, any chance to get this reviewed?

dougwilson commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
-node_modules/
\ No newline at end of file
+node_modules/
+.vscode/
```
I would like to keep this clean for the project. Let's remove this and just keep it in your system-level `.gitignore`.

leewaygroups replied:
Done: Removed the extra lines as suggested.

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
 
 module.exports = function requireAll(options) {
-  var dirname = typeof options === 'string' ? options : options.dirname;
-  var excludeDirs = options.excludeDirs === undefined ? DEFAULT_EXCLUDE_DIR : options.excludeDirs;
```
Please undo all this whitespace changes.

leewaygroups replied:
I sure noticed that the extra whitespaces generated a lot of noise. Sorry about that. I've tried to remove all of them. Hopefully, I caught all.

How about adding an editor config to ensure anyone gets the desired settings?

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+            });
+
+        } else {
+            if (filter instanceof RegExp) {
```
Please use the `toString()` test instead of `instanceof`.

leewaygroups replied:
Fixed.

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+
+                modules[map(match[1], filepath)] = resolve(require(filepath));
+            } else {
+                var match = filter instanceof Function ? filterOutFuncHandler(file, filter) :
```
Please use `typeof` instead of `instanceof`.

leewaygroups replied:
Fixed

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+                modules[map(match[1], filepath)] = resolve(require(filepath));
+            } else {
+                var match = filter instanceof Function ? filterOutFuncHandler(file, filter) :
+                    filter instanceof Array ? filterOutArrayHandler(file, filter) : defaultError();
```
Please use `Array.isArray` instead of `instanceof`

leewaygroups replied:
Fixed

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+                var match = filter instanceof Function ? filterOutFuncHandler(file, filter) :
+                    filter instanceof Array ? filterOutArrayHandler(file, filter) : defaultError();
+                if (match) return;
+                var fileName = file.replace(/\.[^/.]+$/, "")
```
Why is the extension being removed when not using a regular expression? This would be inconsistent and also not let people use custom functions while sniffing for the extension.

dougwilson replied:
It would also cause all dot files to just become an empty string...

leewaygroups replied:
Handling file extensions and determining what the map function gets is now left to the filter function supplied by the user. 

The expectation is that the filter function returns a false value for any file to be excluded and a string filename in the case of files to be included.

leewaygroups replied:
The `ArrayFilterHandler` is still somewhat strict but let me have your comment on it.

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+function filterOutFuncHandler(file, filter) {
+    var result = filter(file);
+    if (typeof result === "boolean") return result;
+    throw new Error('supplied filter function return value must be boolean');
```
Throw `TypeError`s for type errors, please.

leewaygroups replied:
Fixed

dougwilson commented on index.js:
```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
+
+function filterOutFuncHandler(file, filter) {
+    var result = filter(file);
+    if (typeof result === "boolean") return result;
```
Why so strict? In fact, this is a lame interface compared to the regular expression one, even though functions are more flexible. The regular expression one allows you to use capture groups to trim down what `map` gets. Can you make it just let through returning a string as being `true` with a different name to use?

leewaygroups replied:
Fixed as comment earlier.

dougwilson commented on test/test.js:
```diff
@@ -3,212 +3,260 @@ var semver = require('semver');
 
 controllersMap = requireAll({
-  dirname: __dirname + '/controllers',
-  filter: /(.+Controller)\.js$/,
```
Can you undo all these whitespace changes?

leewaygroups commented:
This commit `Further refactor and test following PR review feedback` has all changes addressing the points raised in the PR feedback.  

@dougwilson Kindly review and let me know if I missed something or if there exist some opportunities for improvement.

Cheers!

dougwilson commented:
Hi @leewaygroups thank you so much for putting this PR together and working on it. I'm very sorry it fell off my radar. There were still a couple things, most notable documentation & undoing all those trailing newline removals, but I didn't want to let this go in more circles, so I hope it's OK, but I just amended your PR with the changes and merged 🎉 I did remove the array feature, though, as that can be added back in another PR, but, really, it seems redundant when you can specify a full function.
